doc/Yang Model Guidelines.md

```diff
@@ -0,0 +1,253 @@
+            }
+            leaf IP_TYPE {
+                type xxxx;
+            }
+            leaf PRIORITY {
+                type uint32 {
+                    range xxxx;
+                }
+            }
+```
+
+
+### 3. Hierarchy of the an objects in yang models will be same as for all the objects at same hierarchy in Redis. If any exception is created then it must be recorded properly.
+For Example: 
+
+"Family" of VLAN_INTERFACE and "IP_TYPE" of ACL_RULE should be at same level in yang model too.
+
+#### ABNF
+```
+"VLAN_INTERFACE": {
+        "Vlan100|2a04:f547:45:6709::1/64": {
+            "scope": "global",
+            "family": "IPv6"
+        }
+}
+"ACL_RULE": {
+        "NO-NSW-PACL-V4|DEFAULT_DENY": {
+            "PACKET_ACTION": "DROP",
+            "IP_TYPE": "IPv4ANY",
+            "PRIORITY": "0"
+        }
+}
+```
+####  Yang
+In Yang, "Family" of VLAN_INTERFACE and "IP_TYPE" of ACL_RULE is at same level. 
+```
+container VLAN_INTERFACE {
+        description "VLAN_INTERFACE part of config_db.json";
+        list xxx {
+    ......
+    ......
+            leaf family {
+                type sonic-head:ip-family;
+            }
+        }
+}
+
+container ACL_RULE {
+        description "ACL_RULE part of config_db.json";
+        list xxx {
+        ......
+    ......
+            leaf IP_TYPE {
+                type sonic-head:ip_type;
+            }
+    }
+}
+``` 
+
+### 4.  if an object is used as a key in Redis, it should be a key in yang model, irrespective of need. Exception must be recorded in Yang Model.
+
+Example: VLAN_MEMBER dictionary in ABNF.json has both vlan-id and Port part of the key. So yang model should have the same keys.
+
+#### ABNF
+```
+ "VLAN_MEMBER": {
+        "Vlan100|Ethernet0": {<<<< KEYS
+            "tagging_mode": "untagged"
+        }
+   }
+```
+#### Yang
+```
+container VLAN_MEMBER {
+    description "VLAN_MEMBER part of config_db.json";
+        list xxx {
+            key "vlanid port";<<<< KEYS
+       }
+}
+```
+
+### 5.  It is best to categorized Yang Modules based on a networking components. For Example, it is good to have separate modules for VLAN,  ACL, PORT and IP-ADDRESSES etc.
+```
+sonic-acl.yang
+sonic-interface.yang
+sonic-port.yang
+sonic-vlan.yang
+```
+
+### 6. When it comes about the choice between objects, it is good to stick to standard models. 
+For example: if open-config defines acl-rules as container and have an acl-rules list inside it, then we will stick to the same object types, if a container and list is must in Sonic Yang model.
+Examples:
+```
+container acl-entries {
+    description "Access list entries container";
+    list acl-entry {
+        key "sequence-id";
+```
+
+
+
+### 7.  All must, when, pattern and enumeration constraints will come from .h files (or code). If code has the possibility to have unknown behavior with some config, then we will put a constraint in yang models objects. (This will always be Ongoing effort).
```

Comment:
I would prefer to say "must, when, leafref, pattern and enumeration constraints can be derived from .h files (or code logic)." 
Developer can put any additional constraint to stop invalid configuration (for which there might not be any code logic)

Reply:
AGREE.

Reply:
AGREE